### PR TITLE
'pbsnodes -o' offlines ALL vnodes/compute nodes on a Cray X* series

### PR DESCRIPTION
**Pull request by lisa-altair**
#### Bug/feature Description
* using 'pbsnodes -o' on a Cray X* series causes all of the vnodes to also be marked offline, even when another Mom (that is not offline) was reporting that same vnode.  

#### Affected Platform(s)
* Platform: All

#### Cause / Analysis / Design
* Please see the design: **[project documentation area](https://pbspro.atlassian.net/wiki/spaces/PD/pages/195002450/offline+vnodes+and+pbsnodes+-o+should+only+offline+vnodes+belonging+to+more+than+one+Mom+when+all+the+MoMs+are+offline)**
* Feel free to comment on the discussion on the Forum: **[pbspro community forum](http://community.pbspro.org/t/offline-vnodes-should-only-offline-vnodes-belonging-to-more-than-one-mom-when-all-the-moms-are-offline/864/21)**
* When the pbsnodes -o feature was first designed, it did not account for multiple moms reporting the same set of vnodes.  Thus pbsnodes -o always brought down all vnodes of a Mom.  This behavior was undesirable on a Cray X* series system, because it basically rendered the Cray X* series system unusable once all the compute node vnodes were marked offline.

#### Solution Description
* This pull request changes the behavior of pbsnodes -o to only offline the vnodes that are reported by more than one Mom when all Moms are offline.  As long as one Mom that reports the vnode is not offline, then that vnodes will not be offline.
This more closely aligns with the original intent of pbsnodes -o.

#### Testing logs/output 
[linux_before.txt](https://github.com/PBSPro/pbspro/files/2215681/linux_before.txt)
[linux_after.txt](https://github.com/PBSPro/pbspro/files/2215687/linux_after.txt)
[linux_after_2.txt](https://github.com/PBSPro/pbspro/files/2229665/linux_after_2.txt)
[linux_after3.txt](https://github.com/PBSPro/pbspro/files/2242034/linux_after3.txt)

#### Checklist:

- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [X] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [X] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/server/req_manager.c [hunk 1]
```diff
@@ -2316,10 +2315,34 @@ mgr_node_set(struct batch_request *preq)
...
 				break;	/* all down */
 			pnode = ((mom_svrinfo_t *)(pmom->mi_data))->msr_children[momidx];
+			if ((strcmp(plist->al_name, ATTR_NODE_state) == 0) && (plist->al_op == INCR)) {
+				/* Marking nodes offline.  We should only mark the children vnodes
+				 * as offline if no other mom is up.
+				 */
+				if (pnode->nd_nummoms > 1) {
+					for (imom = 0; imom < pnode->nd_nummoms; ++imom) {
+						mstate = ((mom_svrinfo_t *)(pnode->nd_moms[imom]->mi_data))->msr_state;
+						if (((mstate & INUSE_DOWN) == 0) && ((mstate & INUSE_OFFLINE) == 0)) {
```
Comment (arungrover) [inline src/server/req_manager.c]: Can you please OR both the states into one and then do the check?

File: test/tests/functional/pbs_offline_vnodes.py [hunk 2]
```diff
@@ -104,6 +106,43 @@ def create_multi_vnodes(self, num_moms, num_vnode=3):
             self.server.expect(NODE, {ATTR_NODE_state: 'free'},
                                id=self.moms.values()[i].shortname)
 
+    def verify_other_vnodes_state(self, expected_state):
```
Comment (arungrover) [inline test/tests/functional/pbs_offline_vnodes.py]: Can you please add a doc string for this function. Also, please change the name of the function. Currently it sounds like it is trying to verify some other vnode's state but the function verifies that all the reported vnodes are in expected_state.

File: test/tests/functional/pbs_offline_vnodes.py [hunk 3]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+        self.create_multi_vnodes(2)
+
+        # Offline both of the moms, the vnodes reported by them
+        # will also be offlined
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
+        pbsnodes_cmd = os.path.join(pbs_exec, 'bin', 'pbsnodes')
+        pbsnodes_offline = pbsnodes_cmd + ' -o ' + momA.shortname
+        self.du.run_cmd(self.server.hostname, cmd=pbsnodes_offline, sudo=True)
+
+        pbsnodes_offline = pbsnodes_cmd + ' -o ' + momB.shortname
```
Comment (arungrover) [inline test/tests/functional/pbs_offline_vnodes.py]: you can append both the mom names in the same pbsnodes command.

File: test/tests/functional/pbs_offline_vnodes.py [hunk 4]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+        self.server.expect(
+            NODE, {ATTR_NODE_state: 'offline'},
+            id=momB.shortname, max_attempts=3, interval=2)
+
+        self.verify_other_vnodes_state('offline')
+
+        # Now call pbsnodes -r to clear the offline from MomA
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
+        pbsnodes_cmd = os.path.join(pbs_exec, 'bin', 'pbsnodes')
+        pbsnodes_offline = pbsnodes_cmd + ' -r ' + momA.shortname
```
Comment (arungrover) [inline test/tests/functional/pbs_offline_vnodes.py]: Please change the variable name from pbsnodes_offline to something like pbsnodes_clear_offline or similar

File: test/tests/functional/pbs_offline_vnodes.py [hunk 5]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+            id=momA.shortname, max_attempts=3, interval=2)
+
+        self.server.expect(
+            NODE, {ATTR_NODE_state: 'offline'},
+            id=momB.shortname, max_attempts=3, interval=2)
+
+        self.verify_other_vnodes_state('offline')
+
+        # Now call pbsnodes -r to clear the offline from MomA
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
```
Comment (arungrover) [inline test/tests/functional/pbs_offline_vnodes.py]: You already have pbsnodes_cmd made at this point, you don't have to recreate it

File: src/server/req_manager.c (hunk 1 above)
Comment (lisa-altair) [reply src/server/req_manager.c]: Okay.

File: test/tests/functional/pbs_offline_vnodes.py (hunk 2 above)
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: How about "verify_vnodes_state"?  I will add a doc string.

File: test/tests/functional/pbs_offline_vnodes.py (hunk 3 above)
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: Cool!  Will do.

File: test/tests/functional/pbs_offline_vnodes.py (hunk 4 above)
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: Okay

File: test/tests/functional/pbs_offline_vnodes.py (hunk 5 above)
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: You're right!

File: test/tests/functional/pbs_offline_vnodes.py [hunk 6]
```diff
@@ -374,10 +355,151 @@ def test_fail_action_startup_hook(self):
...
+        self.create_multi_vnodes(1)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=single_mom.shortname, max_attempts=3,
+                           interval=2)
+
+        mom_host = single_mom.shortname
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
+        pbsnodes_cmd = os.path.join(pbs_exec, 'bin', 'pbsnodes')
+        pbsnodes_offline = pbsnodes_cmd + ' -o ' + mom_host
+        self.du.run_cmd(self.server.hostname, cmd=pbsnodes_offline, sudo=True)
```
Comment (vccardenas) [inline test/tests/functional/pbs_offline_vnodes.py]: run_cmd() works better with a list of arguments, not a string. Pass pbsnodes_offline = [pbsnodes_cmd, '-o', mom_host]

File: test/tests/functional/pbs_offline_vnodes.py [hunk 7]
```diff
@@ -374,10 +355,151 @@ def test_fail_action_startup_hook(self):
...
+            momB.delete_vnode_defs()
+
+        self.create_multi_vnodes(2)
+
+        # Offline only one of the moms, the other mom and her children
+        # should still be free
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
+        pbsnodes_cmd = os.path.join(pbs_exec, 'bin', 'pbsnodes')
+        pbsnodes_offline = pbsnodes_cmd + ' -o ' + momA.shortname
+        self.du.run_cmd(self.server.hostname, cmd=pbsnodes_offline, sudo=True)
```
Comment (vccardenas) [inline test/tests/functional/pbs_offline_vnodes.py]: run_cmd() works better with a list of arguments, not a string. Pass pbsnodes_offline = [pbsnodes_cmd, '-o', momA.shortname]

File: test/tests/functional/pbs_offline_vnodes.py [hunk 8]
```diff
@@ -374,10 +355,151 @@ def test_fail_action_startup_hook(self):
...
+            momB.delete_vnode_defs()
+
+        self.create_multi_vnodes(2)
+
+        # Offline both of the moms, the vnodes reported by them
+        # will also be offlined
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
+        pbsnodes_cmd = os.path.join(pbs_exec, 'bin', 'pbsnodes')
+        pbsnodes_offline = pbsnodes_cmd + ' -o ' + momA.shortname + momB.shortname
+        self.du.run_cmd(self.server.hostname, cmd=pbsnodes_offline, sudo=True)
```
Comment (vccardenas) [inline test/tests/functional/pbs_offline_vnodes.py]: run_cmd() works better with a list of arguments, not a string. Pass pbsnodes_offline = [pbsnodes_cmd, '-o', momA.shortname, momB.shortname]

File: test/tests/functional/pbs_offline_vnodes.py [hunk 9]
```diff
@@ -374,10 +355,151 @@ def test_fail_action_startup_hook(self):
...
+
+        self.server.expect(
+            NODE, {ATTR_NODE_state: 'offline'},
+            id=momB.shortname, max_attempts=3, interval=2)
+
+        self.verify_other_vnodes_state('offline')
+
+        # Now call pbsnodes -r to clear the offline from MomA
+        pbsnodes_clear_offline = pbsnodes_cmd + ' -r ' + momA.shortname
+        self.du.run_cmd(self.server.hostname, cmd=pbsnodes_clear_offline, sudo=True)
```
Comment (vccardenas) [inline test/tests/functional/pbs_offline_vnodes.py]: run_cmd() works better with a list of arguments, not a string. Pass pbsnodes_clear_offline = [pbsnodes_cmd, '-r', momA.shortname]

File: test/tests/functional/pbs_offline_vnodes.py [hunk 10]
```diff
@@ -104,6 +106,46 @@ def create_multi_vnodes(self, num_moms, num_vnode=3):
             self.server.expect(NODE, {ATTR_NODE_state: 'free'},
                                id=self.moms.values()[i].shortname)
 
+    def verify_vnodes_state(self, expected_state):
```
Comment (vccardenas) [inline test/tests/functional/pbs_offline_vnodes.py]: Change this to "verify_other_vnodes_state" instead of "verify_vnodes_state".

Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: That's what it used to be.  Arun asked for it to be changed.  The new name is verify_vnodes_state.  Is there some problem with the current name?

Comment (vccardenas) [reply test/tests/functional/pbs_offline_vnodes.py]: I should have said that verify_vnodes_state() is not being called anywhere.

Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: Ahh.  Thank you!  It is now being called.  :)

File: test/tests/functional/pbs_offline_vnodes.py (hunk 6 above)
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: ok.  will update them.

**Comment by lisa-altair** [comment]
Thank you @arungrover.  I have attached my recent test logs (linux_after_2.txt).

**Comment by vccardenas** [comment]
@lisa-altair, the PTL tests look good to me.

**Comment by arungrover** [comment]
@lisa-altair please rebase and squash your changes

**Comment by arungrover** [comment]
I just realize that Voltaire reviewed only PTL test. You need another reviewer to review the C code.

File: test/tests/functional/pbs_offline_vnodes.py [hunk 11]
```diff
@@ -137,21 +179,7 @@ def test_single_mom_hook_failure_affects_vnode(self):
...
-            # is offline
-            for v1 in vlist:
-                if vname is None:
-                    vname = v1
-                # Check that the vnode is in offline state
-                self.server.expect(
-                    VNODE, {'state': 'offline'}, id=v1, max_attempts=3,
-                    interval=2)
-        elif cpuset_mom is True:
+        if cpuset_mom is True:
```
Comment (bhroam) [inline test/tests/functional/pbs_offline_vnodes.py]: I think the cpuset chunk of code can move up to your function.  You need to create a list from what status() returns.  This is easily done via list comprehension.  You do it by: vlist = [x['id'] for x in vnl if x['id'] != single_mom.shortname]
If you want to know what that black magic is, I'll walk you through it on a whiteboard.

File: test/tests/functional/pbs_offline_vnodes.py [hunk 12]
```diff
@@ -489,8 +489,9 @@ def test_pbsnodes_multi_mom_offline_online(self):
         self.verify_vnodes_state('offline')
 
         # Now call pbsnodes -r to clear the offline from MomA
-        pbsnodes_clear_offline = [pbsnodes_cmd,'-r', momA.shortname]
-        self.du.run_cmd(self.server.hostname, cmd=pbsnodes_clear_offline, sudo=True)
+        pbsnodes_clear_offline = [pbsnodes_cmd, '-r', momA.shortname]
+        self.du.run_cmd(self.server.hostname, cmd=pbsnodes_clear_offline,
+                        sudo=True)
```
Comment (bhroam) [inline test/tests/functional/pbs_offline_vnodes.py]: Do you need to run this as sudo?  You are added as a manager by PTL.  That should be sufficient.

File: src/server/req_manager.c [hunk 13]
```diff
@@ -2316,10 +2315,34 @@ mgr_node_set(struct batch_request *preq)
 		if (numnodes == 1)
 			break;	/* just the one vnode */
 		else if (preq->rq_ind.rq_manager.rq_objtype == MGR_OBJ_HOST) {
+			int imom;
+			int update_mom_only = 0;
+			unsigned long mstate;
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply src/server/req_manager.c]: I think this was moved in an earlier commit.  I don't see what you are referring to...

File: src/server/req_manager.c [hunk 14]
```diff
@@ -2186,6 +2186,7 @@ mgr_node_set(struct batch_request *preq)
 
 	} else if (preq->rq_ind.rq_manager.rq_objtype == MGR_OBJ_HOST) {
 		/* Operating on all vnodes on a named host          */
+		/* if it is the last/only host			    */
 		/* find the mom and get the first vnode in her list */
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply src/server/req_manager.c]: fixed.

File: src/server/req_manager.c [hunk 15]
```diff
@@ -2316,10 +2315,34 @@ mgr_node_set(struct batch_request *preq)
...
+			int update_mom_only = 0;
+			unsigned long mstate;
+
 			/* next vnode under the Mom */
 			if (++momidx >= ((mom_svrinfo_t *)(pmom->mi_data))->msr_numvnds)
 				break;	/* all down */
 			pnode = ((mom_svrinfo_t *)(pmom->mi_data))->msr_children[momidx];
+			if ((strcmp(plist->al_name, ATTR_NODE_state) == 0) && (plist->al_op == INCR)) {
+				/* Marking nodes offline.  We should only mark the children vnodes
+				 * as offline if no other mom is up.
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply src/server/req_manager.c]: ok

File: test/tests/functional/pbs_offline_vnodes.py [hunk 16]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
-            self.server.manager(MGR_CMD_CREATE, NODE,
-                                id=m.shortname)
+    def test_pbsnodes_o_single_mom(self):
+        """
+        Offline a mom using pbsnodes -o.
+        Since it is the only mom, all vnodes reported by her
+        should also be offline.
+        """
+        single_mom = ""
+        single_mom = self.moms.values()[0]
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: I copied it from an existing test in this suite.  I have removed them both.

File: test/tests/functional/pbs_offline_vnodes.py [hunk 17]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+
+    def test_pbsnodes_o_multi_mom_only_one_offline(self):
+        """
+        Offline one mom using pbsnodes -o.
+        In the case of multiple moms reporting the same set of vnodes,
+        none of the vnodes should be marked offline,
+        including the children vnodes.
+        NOTE: This test needs moms to report the same set of vnodes.
+        """
+        if len(self.moms.values()) != 2:
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: Okay, but it matches all other instances of this call in this file...

File: test/tests/functional/pbs_offline_vnodes.py [hunk 18]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+            self.skipTest("Provide 2 moms while invoking test")
+
+        if self.moms.values()[0].is_cpuset_mom() is True or self.moms.values()[
+                1].is_cpuset_mom() is True:
+            self.skipTest("Skipping test on cpuset moms")
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: This matches an existing test.  I'd like to  leave it as is.

File: test/tests/functional/pbs_offline_vnodes.py [hunk 19]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+        if momB.has_vnode_defs():
+            momB.delete_vnode_defs()
+
+        self.create_multi_vnodes(2)
+
+        # Offline only one of the moms, the other mom and her children
+        # should still be free
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
+        pbsnodes_cmd = os.path.join(pbs_exec, 'bin', 'pbsnodes')
+        pbsnodes_offline = pbsnodes_cmd + ' -o ' + momA.shortname
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: I'm pretty sure this was fixed in an earlier commit.  Are you looking at the latest changes?

File: test/tests/functional/pbs_offline_vnodes.py [hunk 20]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+        if momA.has_vnode_defs():
+            momA.delete_vnode_defs()
+        if momB.has_vnode_defs():
+            momB.delete_vnode_defs()
+
+        self.create_multi_vnodes(2)
+
+        # Offline only one of the moms, the other mom and her children
+        # should still be free
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: Yes, it helps keep the line from being too long.

File: test/tests/functional/pbs_offline_vnodes.py [hunk 21]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+            self.skipTest("Provide 2 moms while invoking test")
+
+        if self.moms.values()[0].is_cpuset_mom() is True or self.moms.values()[
+                1].is_cpuset_mom() is True:
+            self.skipTest("Skipping test on cpuset moms")
+
+        momA = ""
+        momA = self.moms.values()[0]
+        momB = ""
+        momB = self.moms.values()[1]
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: fixed

File: test/tests/functional/pbs_offline_vnodes.py [hunk 22]
```diff
@@ -374,10 +352,156 @@ def test_fail_action_startup_hook(self):
...
+        """
+        Offline a mom using pbsnodes -o.
+        Since it is the only mom, all vnodes reported by her
+        should also be offline.
+        """
+        single_mom = ""
+        single_mom = self.moms.values()[0]
+        self.create_multi_vnodes(1)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=single_mom.shortname, max_attempts=3,
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply test/tests/functional/pbs_offline_vnodes.py]: removed all max_attempts

File: src/server/req_manager.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (bhroam) [reply src/server/req_manager.c]: It looks fine now.  This was a comment based on the first commit.

**Comment by lisa-altair** [comment]
I have attached the latest test run results.  @bhroam and @arungrover Please have a look at the code and test changes.

**Comment by lisa-altair** [comment]
I have rebased and squashed, please take it forward.

**Comment by lisa-altair** [comment]
Thank you!

**Pull request merged**